heartbeat/scheduler/scheduler.go

```diff
@@ -242,65 +243,94 @@ func (s *Scheduler) runOnce(runAt time.Time, taskFn timerqueue.TimerTaskFn) {
+// The wait group passed into this function expects to already have its count incremented by one.
+func (sj *schedJob) run() (startedAt time.Time) {
+	sj.wg.Add(1)
+	sj.activeTasks.Inc()
+	if sj.jobLimitSem != nil {
+		sj.jobLimitSem.Acquire(sj.ctx, 1)
+	}
+
+	startedAt = sj.runTask(sj.entrypoint)
+
+	sj.wg.Wait()
 	return startedAt
 }
 
 // runRecursiveTask runs an individual task and its continuations until none are left with as much parallelism as possible.
 // Since task funcs can emit continuations recursively we need a function to execute
 // recursively.
 // The wait group passed into this function expects to already have its count incremented by one.
-func (s *Scheduler) runRecursiveTask(jobCtx context.Context, task TaskFunc, wg *sync.WaitGroup, jobSem *semaphore.Weighted) (startedAt time.Time) {
-	defer wg.Done()
+func (sj *schedJob) runTask(task TaskFunc) time.Time {
+	defer sj.wg.Done()
+	defer sj.activeTasks.Dec()
 
 	// The accounting for waiting/active tasks is done using atomics.
 	// Absolute accuracy is not critical here so the gap between modifying waitingTasks and activeJobs is acceptable.
-	s.stats.waitingTasks.Inc()
+	sj.scheduler.stats.waitingTasks.Inc()
 
 	// Acquire an execution slot in keeping with heartbeat.scheduler.limit
 	// this should block until resources are available.
 	// In the case where the semaphore has free resources immediately
 	// it will not block and will not check the cancelled status of the
 	// context, which is OK, because we check it later anyway.
-	limitErr := s.limitSem.Acquire(jobCtx, 1)
-	s.stats.waitingTasks.Dec()
+	limitErr := sj.scheduler.limitSem.Acquire(sj.ctx, 1)
+	sj.scheduler.stats.waitingTasks.Dec()
 	if limitErr == nil {
-		defer s.limitSem.Release(1)
+		defer sj.scheduler.limitSem.Release(1)
 	}
 
 	// Record the time this task started now that we have a resource to execute with
-	startedAt = time.Now()
+	startedAt := time.Now()
 
 	// Check if the scheduler has been shut down. If so, exit early
 	select {
-	case <-jobCtx.Done():
+	case <-sj.ctx.Done():
 		return startedAt
 	default:
-		s.stats.activeTasks.Inc()
+		sj.scheduler.stats.activeTasks.Inc()
 
-		continuations := task(jobCtx)
-		s.stats.activeTasks.Dec()
+		continuations := task(sj.ctx)
+		sj.scheduler.stats.activeTasks.Dec()
 
-		wg.Add(len(continuations))
+		sj.wg.Add(len(continuations))
+		sj.activeTasks.Add(len(continuations))
 		for _, cont := range continuations {
 			// Run continuations in parallel, note that these each will acquire their own slots
 			// We can discard the started at times for continuations as those are
 			// irrelevant
-			go s.runRecursiveTask(jobCtx, cont, wg, jobSem)
+			go sj.runTask(cont)
 		}
-		if jobSem != nil && len(continuations) == 0 {
-			jobSem.Release(1)
+		// There is always at least 1 task (the current one), if that's all, then we know
+		// there are no other jobs active or pending, and we can release the jobLimitSem
+		if sj.jobLimitSem != nil && sj.activeTasks.Load() == 1 {
```

Comment:
This is the core of the fix, being able to count the actual number of active tasks